### DISCUSSION TITLE
Text Expander Plugin

### DISCUSSION THREAD

**Discussion by mbluett88** [zim-desktop-wiki/zim-desktop-wiki#2547] [Development]
I am developing a Text Expander plugin that can be inserted into any page. It is almost complete.
I am saving the contents of each Expander buffer as a separate page so as to not make the page that holds the Expander objects too large. At the moment, I have the Expander pages stored in a separate Expander tree item/folder under the root Notebooks. My preference would be to make it so this Expander group is not part of the Notebooks tree because the pages are not like a typical Zim page (i.e., these Expander pages do not have a title nor a creation date). Although I could leave it as it is.

I would like to get some feedback as to which would be the preferable method of storing these pages. I would prefer not to contaminate the Zim Notebooks tree. However, having them as part of the Zim tree means that they can easily be searched using the Zim search mechanism. Maybe I could somehow make the Expander tree item invisible. I have not researched if that is possible.

Comments?

**Comment by Simplenuity** [comment]
Hi @mbluett88 

Thanks for your efforts!

What do you mean by "Text Expander plugin that can be inserted into any page"? Isn't a text expander inserting previously defined text blocks into a page by using keywords/key combinations? Which would result the page itself looking like the text block has been typed in manually?

If I understand the concept correctly, I can think of two suitable locations (Linux) for a TextExpander sub folder:
* `~/.config/zim/`  
* `~/.local/share/zim/`

with the second one preferable due to its nature.

**Reply by mbluett88** [reply]
Well, I could use those folders. The only problem would be that the builtin Zim search function would not search those files. That is a bit of a problem for me, at least. The main reason for me using Zim was to find an app that I could quickly and easily search through my notes.

**Comment by jtoppine** [comment]
Is the subobjects being part of the main page making the pages too large (computer memory/processing -wise) actually a real world problem?

Asking because I would think Zim can handle quite large pages well, and in my mind the possibility to expand/contract content is more of a visual aid to help human readers to navigate in a big page - not to chop up the file into multiple parts to make it easier to handle for computer.

FWIW, the fact that Zim files are just glorified text files than can be almost as well read and edited by any text based viewer is the main reason I use Zim. Big part of the appeal of Zim is the simple, standard document structure which can be used and edited through Zim, but also easily accessed and processed by custom tools like website generators, cloud sync software, and whatever contraptions users might have.

For these tools I would think it would be preferable to have the expander content inside the main document, to keep parsing logic simple (tools can just ignore unrecognized markup, but would need complex logic to navigate an additional subdocument tree). 

Even for manual handling of notes files, one might want to take a note file from zim data folder and copy it elsewhere for whatever use. It would be confusing to have additional sub-document directory tree to worry about, as not including all the related files would mess up the internal structure of the content.

**Reply by Simplenuity** [reply]
@jtoppine, your description sounds more like foldable text blocks rather than text expander?
At first I also assumed it's about foldable text blocks till DuckDuckGo told me otherwise.
Curious now what we're actually talking about :)

Also, for the discussion it might be helpful if @mbluett88 could provide some examples?

**Reply by jtoppine** [reply]
@Simplenuity Yeah I was, or am, assuming this is about foldable text blocks. Not sure though!

**Comment by mbluett88** [comment]
> Is the subobjects being part of the main page making the pages too large (computer memory/processing -wise) actually a real world problem?

For some, it still is. Came across a good assessment of memory usage before, which, unfortunately, I did not keep a copy of. Basically, it stated that on an 8 Gig system, roughly no more than 4 Gig gets used for buffer space. So, if you consider multiple apps running simultaneously each using some of that 4 Gig, it could result in a problem. For example, I use Blender from time-to-time. That can use a large amount of memory. Currently, I only have an 8 Gig system. I am rethinking that due to Blender. But making notes at the same time I am using Blender could be a problem.

Also, Python is interpretative. That means it is slower than a compiled program. I don't know if you have examined what happens behind the scenes with Zim or not. But, when you do examine it you can see the enormous and sometimes excessive processing that goes on. So, as documents grow large in size they can result in Zim performing much slower.

These are all things I think about when programming. I am old school, when it was necessary to deal with Megs of memory and not Gigs. And, of course, much slower CPUs and disks.

> in my mind the possibility to expand/contract content is more of a visual aid to help human readers to navigate in a big page - not to chop up the file into multiple parts to make it easier to handle for computer.

You are right: That is what it is intended for. The visual part is an aid; but, the processing behind the scenes is important for the sake of resource usage. A page could be home to multiple collapsible sections, each, potentially, containing a lot of text. The idea is that as you expand a section the text is read into memory and then removed from memory when collapsed. With today's SSDs that action is very fast. Maybe I am being overly concerned. But, I am not certain of that.

> For these tools I would think it would be preferable to have the expander content inside the main document, to keep parsing logic simple (tools can just ignore unrecognized markup, but would need complex logic to navigate an additional subdocument tree).

Very good point, that had not occurred to me. I am now quite glad I posted my original question.

> your description sounds more like foldable text blocks rather than text expander? 
> At first I also assumed it's about foldable text blocks till DuckDuckGo told me otherwise.
Curious now what we're actually talking about 

Yes, foldable text is another way of describing it. GTK uses the term Expander so I kept to that. Also, the term Collapsible is good. I don't think the name is that important as long as it is accompanied by an explanation, as you so rightly pointed out, would have been helpful.

If you would prefer I can provide a more concise description. But, I think after this discussion it is probably clear what the intent is.

**Reply by jtoppine** [reply]
Ok it seems clear what we are talking about :)

About memory/cpu usage; generally text is orders of magnitude more lightweight in the sense that you can fit incredible amounts of text in the same memory space as a small bitmap image for example. All of the bible (english) encoded in UTF-8 is around 4MB (thousand bibles to fill 4Gb in buffers) whereas a single screenshot/framegrab of a FullHD screen as uncompressed RGB bitmap is around 6MB. Blender by nature is an app that really can use up a lot memory, multiple GB's easily, in amounts that any collection of books or a note taking app would absolutely pale in comparison.

That is why I was sceptical of whether the amount of text in a single document can ever realistically be a noticeable performance concern in comparison to all of the other goings on in the executable, even more so the OS. Of course the big question is that memory impact of plain text can be multiplied by whatever extra in-memory working details Zim, or GTK, and/or Python needs to do allocate to actually display that amount of text in an editable control. That multiplier I have zero idea of, and it can potentially be significant. I'd guess it's not in this case, but I might as well be wrong! (that is also why I formulated my initial response as a question)

My personal experience is that Zim is already pretty snappy and lightweight even with large documents, but my notes tend to not be _that_ large so idk! Just afraid optimizing for unloading/loading subdocuments might be a lot of work for perhaps even not noticeable difference (as well as the desire to keep the underlying data structure simple for external tools and ability to treat notes as glorified *.txt files).

**Comment by Simplenuity** [comment]
Thanks, @mbluett88, for the clarification and thanks, @jtoppine, for getting me on the right track! So my first intuition has been correct...

The following I typed up assuming we're talking abut collapsible text blocks and before getting misguided by my search:

------------
Do I understand correctly that there would be four solutions possible saving the content?

* in the file (page) itself the expander element is used in

* in a separate file that will be saved in the same location the page is saved, or in a sub folder to this page

* in a separate file that will be save in a central location inside the notebook

* in a separate file outside the notebook

The first aspect that comes to my mind and is also the most important to me:

**Will the notebook still be usable without Zim, i.e. by using a normal text editor?**

This would be crucial for me, for example to be able to access and use the notebook in Markor on an Android device. Or to use notebooks if I'm working on the console and/or remotely.

Considering this aspect I obviously would prefer to have the Text Expander plugin inserting the element/code **directly into the Zim page** where it's supposed to be rendered. Like it is with code blocks and tables. 

If just too much code, so a convenient usage of the page is hardly possible, I could live with a solution similar to pictures. A link to the Text Expander element is inserted in the Zim page while the element itself is saved in the **attachment folder**.

------------

**Comment by introt** [comment]
If I've understood correctly, by an "expander" you mean a collapsible element similar to the HTML `<details><summary>`-tag pair, demoed below:

<details><summary>The summary is always visible</summary>while the details can be expanded and folded with a click.

<details><summary><h3>The summary can also be</h3></summary>
<ul>
<li>a header</li>
<li><details><summary>a list item</summary><ul><li>these need some CSS to look nice</li><li>see #1667#discussioncomment-1709219 for a minimal example</li></ul></details></li>
<li>etc</li>
</ul>
</details>
</details>

> I am saving the contents of each Expander buffer as a separate page so as to not make the page that holds the Expander objects too large.

Given our file-based storage, I'd split the problem into two: collapsible blocks[^blocks] and transclusion[^transclusion]. Doing so, the first would allow the user to navigate long pages quicker by collapsing headers and/or long lists as well as enchant outlining and active recall workflows without introducing new markup nor affecting the ability to edit the source files outside Zim. Transclusion - which you seem to already have suitable code for - not only allows splitting super long pages into multiple files, but enables the user to reuse those shorter pages as part of multiple longer ones.[^obs]

[^blocks]: previously discussed over at #1667 as well as others

[^transclusion]: previously discussed at length here #1818, see also #1706 and #1907. Somewhat related, back in 2013 there was also the experimental SideNotePlugin developed by Jaap, which added "an extra widget with a text buffer in a seperate pane -- to keep notes for a specific page that one wants to keep outside of the main part of the page. It saved those notes in a plain-text file under the main page's attachment directory.

[^obs]: Both features exist in many "competitors", including the very-likely-Zim-inspired Obsidian which uses the terms [folding](https://help.obsidian.md/Editing+and+formatting/Folding) and [embedding](https://help.obsidian.md/Linking+notes+and+files/Embed+files). One reason to "return the favor"[^favor] is to facilitate eventual interoperability once we get around to supporting markdown (#26).

[^favor]: and I very much mean this in the best way, we're very much on the same (stored-in-a-local-text-file) page here: "Obsidian uses open, non-proprietary files, so you're never locked in, and can preserve your data for the long term."

For exports, I'd suggest using the aforementioned HTML tags for the collapsible blocks and including the transcluded content in full wherever used to enable javascript-free standalone files.

**Reply by introt** [reply]
> My preference would be to make it so this Expander group is not part of the Notebooks tree because the pages are not like a typical Zim page (i.e., these Expander pages do not have a title nor a creation date).

Apart from "that's not how the others do it" (nor does Obsidian allow editing the transcluded pages directly afaik), the user-visible titles and creation dates are part of the default page template, not a required property. If your desired flow is more like inserting a code block than moving existing text to a new page, I'd suggest letting the user define where they want the new page stored (eg. as a child or a sibling, or in a top-level branch of the user's choosing) with defaults editable in plugin preferences, including which template to use for the "expander" pages.

**Reply by mbluett88** [reply]
Since I am not educated as a Computer Scientist, the term Transclusion was not part of my vocabulary. I look it up and, now, understand it's significance.

**Comment by mbluett88** [comment]
jtoppine has previously stated the following which made sense to me.

> For these tools I would think it would be preferable to have the expander content inside the main document, to keep parsing logic simple (tools can just ignore unrecognized markup, but would need complex logic to navigate an additional subdocument tree).

For that reason I have dropped back to a previous version which does not save the Expander/Collapser content as separate pages.

**Reply by introt** [reply]
Yes, I agree that keeping the expanding content inside the main document is a good default - this way no new markup is needed for eg. collapsible headers, and there aren't any loops to account for. Could you however publish the code for the other version as well? I'd like to give making a transclusion plugin a shot and your code sounds to already be mostly there, loading buffers from separate files.

**Reply by mbluett88** [reply]
Yes, I can provide that code as well. It is not totally functional but does include the functional part of loading files and saving them.

I am not sure what the best way is to supply this code. Perhaps you can help with that.

I am back to finishing the 'normal' version and have almost got it where I think it is ready for release.

Since the Expander plugin relies on much of the formatting capabilities of the PageView class, I made an attempt to instantiate PageView within my plugin. PageView makes use of a ScrolledWindow. I was having trouble trying to find a way to automatically resize the Expander window as text is added. Using the Gtk.EventBox does not have that problem. So, that has forced me to 'reinvent' the wheel so to speak. I have copied some of the PageView code into my Expander plugin. Not ideal, but a solution.

I would like to get the Expander buffer to take advantage of the formatting options in the menu. Currently, it cannot do that. I am struggling to get my head around how Zim does that. I have a trace of the calls that are made, but my coding attempts to follow that scenario have so far been unsuccessful. If you have any ideas/insights I would be most grateful.

The Zim TextBuffer.update_editmode() method results in the PageView.do_textstyle_changed() method being called. This does not happen because the 'update_editmode()' method fails to get any tags (tags = self.iter_get_zim_tags(iter)) even though there is a valid 'iter'.

The 'iter' it is using is after I have clicked in the middle of some bold-faced text.

Onwards!!

**Reply by introt** [reply]
> I am not sure what the best way is to supply this code. Perhaps you can help with that.

For one-offs there's https://gist.github.com/ (like pastebin, but supports multiple files and versions), otherwise you can add a branch to the plugin's main git repo.

> Since the Expander plugin relies on much of the formatting capabilities of the PageView class, I made an attempt to instantiate PageView within my plugin.

I'm afraid I can't really follow along without seeing the code. Which capabilities do you need that implementing a `PageViewExtension` doesn't offer?

I'm also a bit perplexed as to why the Expander has its own window - the separate buffer made sense as their contents came from other pages - have you looked into hidden text ranges? Do they not fit the bill?

Happy hacking!

**Reply by mbluett88** [reply]
Yes, I will look closer at PageViewExtensions. 

> Introit:  I'm also a bit perplexed as to why the Expander has its own window

This is probably my incorrect use of the word "window" in the context of a Zim page. What I am actually referring to is a text box (within a Zim page window).

As to "hidden text ranges", what would that accomplish that an Gtk.Expander object does not accomplish? Or, are you referring to my older idea of separate files for each Expander object?

**Reply by mbluett88** [reply]
So, to instantiate the PageViewExtension class I would need to do something like this (within my ExpanderWidget class):

``
window = Window()
navigation = NavigationModel(window)
self.pageview = PageView(notebook, navigation)
self.pageview_extension = PageViewExtension(plugin, self.pageview)
``

It is not clear why I should use PageViewExtension when PageView is required as a parameter to PageViewExtension (in order to instantiate it).

Why not just use PageView as I don't need access to any of the methods within the PageViewExtension class?

**Reply by introt** [reply]
> As to "hidden text ranges", what would that accomplish that an Gtk.Expander object does not accomplish? Or, are you referring to my older idea of separate files for each Expander object?

I'm referring to your newer idea; with "hidden text ranges" (which I'm under the impression are what eg. https://github.com/geany/geany uses for collapsing classes and functions) you don't need to re-implement any editor functionality as you can use the existing text buffer.

> It is not clear why I should use PageViewExtension when PageView is required as a parameter to PageViewExtension (in order to instantiate it).

I get the impression that you haven't read https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/develop/PLUGIN_WRITING.md - the plugin classes are instantiated for you when the plugin is loaded:

> When you define a subclass of such an extension class, it will be loaded automatically by the plugin for each instance of the target component.

When your plugin is enabled and gets loaded, it gets the existing PageView instance as a parameter - you don't need to instantiate your own.

Hope this helps clear things up!

**Reply by mbluett88** [reply]
> I'm referring to your newer idea; with "hidden text ranges" (which I'm under the impression are what eg. https://github.com/geany/geany uses for collapsing classes and functions) you don't need to re-implement any editor functionality as you can use the existing text buffer.

I think I need to explain things more clearly about my implementation.

I use the InsertedObjects class to instantiate my Expander object. I tried using the PageView approach that other plugins use; but, that did not result in what I needed. I have gone through many trials of different options and InsertedObjects functions the best for me. 

Because I use that approach my plugin does not get the parameters automatically filled in when I use PageView.

You see, I need to use both these approaches. Not just one or the other. There are no examples of doing this within the existing set of plugins. So, it has been a difficult and very experimental road. But, I am still learning more things about Zim in the process. I find the Zim implementation to be quite complex. Possibly more complex than it needs to be.

I tried mimicking the CodeView plugin. That plugin does not perform any formatting as is done in Zim (e.g.,  Ctrl-b). So, I was forced to implement my own code which now supports some of the formatting approach that Zim uses. There is still some things that I have not yet tested that probably don't work, like numbered bullets.

I am trying to have the Expander object work as close to the way Zim works as I can. It needs to act as if it is seamless. That is why I am currently trying to get the menu formatting buttons to work the same way they do outside of my Expander object.

For more clarity, when you left-click in the middle of a section of bolded text I want the bold menu item to be selected, as it is outside of my Expander object. That functionality appears to be part of the PageView class.

**Reply by introt** [reply]
I'm afraid I'm going to need to see the code or the Expander object in action to understand what it is/does. My current understanding is that you're using the same Gtk.Expander as used in the tree view, pictured below...

![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/26069787/ae3ed1e0-3ddd-4495-8733-5d8fcf2618d8)

...to achieve expandable text, which is in my understanding in some ways similar to what's done using hidden ranges here:

![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/26069787/6ee876c0-9692-45f7-ac62-88d41413757e)

I think I have some idea of what your current implementation does, but I'm not sure whether recreating the PageView is the way to do it - you're going to incur massive technical depth and loose any plugin functionality (including tables). If you want to go ahead with an InsertedObject/"separate text box" I'd suggest looking into how opening pages in a new window works and see if you can cram a fully functional page view into another using a similar approach - you'll then get the seamlessness by "simply" making sure that the signals are handled from/by the currently active "text box".

**Reply by introt** [reply]
I'd also like to add (re:collapsible headers + in general) that a sane implementation may require (extensive) changes to Zim's side of the equation, which is preferable to trying to do everything on the plugin's side - expanding/collapsing text is a pretty fundamental editor feature so the maintainability of the code is pretty paramount to its introduction.

**Comment by mbluett88** [comment]
> Gtk.Expander as used in the tree view, pictured below...

Yes, that is correct and it has it's own buffer as created using the InsertedObjectTypeExtension
which I have to use in order to differentiate between the Zim Page PageView buffer and that within my Expander object.

I cannot use the PageView class, as do some other plugins, as it, currently, has no way of knowing anything about the buffer content within the Expander object. This is, for example, why the CodeView plugin uses the InsertedObjectTypeExtension class.

This class provides two methods that are very important when updating the Zim Page buffer with information from my Expander object buffer (or that of the CodeView object buffer): `model_from_data()` and `data_from_model()`.

> I'd suggest looking into how opening pages in a new window works and see if you can cram a fully functional page view into another using a similar approach - you'll then get the seamlessness by "simply" making sure that the signals are handled from/by the currently active "text box".

The problem with that is that there are two separate buffers involved. So, that approach would never work. The reason is that PageView does not, currently, have the ability to process any formatted text contained within an object. So, changing the Zim base code would be necessary to make that happen.

The Zim base code may have to be changed to handle such a plugin as I am trying to produce. I was trying, desperately, not to do that because it may complicate things. But, I am not sure it is possible to produce a well-designed plugin without modifying the base code.

For example, someone was forced to modify the base Zim code to properly incorporate Tables which are also considered a plugin.

As to the code I have, you are certainly welcome to it, but I don't know how to get it to you.

I could create a new Github entry, I guess, under Zim plugins and put the code there. But, it is, in my opinion, not ready for primetime.

**Reply by introt** [reply]
>  it has it's own buffer [...] which I have to use in order to differentiate between the Zim Page PageView buffer and that within my Expander object.

Do you mean visually? Mayhaps you could post some screenshots of what you have and/or are working towards.

> The problem with that is that there are two separate buffers involved [...] So, changing the Zim base code would be necessary to make that happen. [...] I was trying, desperately, not to do that because it may complicate things.

Trying to work around Zim code base's limitations will not only produce a non-well-designed plugin, but either make it unfeasible to include in the default plugin set due to being hard to maintain, or make it undependable in the eyes of the users due to being constantly broken with updates. As I've typed before, - "expanding/collapsing text is a pretty fundamental editor feature so the maintainability of the code is pretty paramount to its introduction."

> As to the code I have, you are certainly welcome to it, but I don't know how to get it to you.
>
> I could create a new Github entry, I guess, under Zim plugins and put the code there. But, it is, in my opinion, not ready for primetime.

Don't worry, being on GitHub doesn't equate to being ready for primetime - take my numerous draft pull requests on this repo for example.

**Reply by mbluett88** [reply]
Hola Introt,

I came up with a version of  the Expander that is integrated directly within the Zim base code (as per your suggestion).

I have published it at https://github.com/mbluett88/zim-desktop-wiki/tree/develop

I am new to Github. I think it is setup correctly. But, I have no idea how Jaap will know that I have published it. Is it somehow automatic after I forked it?

**Reply by introt** [reply]
Hi, mbluett, I'll take it out for a spin! To let Jaap know, you can open a pull request here: https://github.com/zim-desktop-wiki/zim-desktop-wiki/pulls

**Reply by introt** [reply]
Oh, I'm liking this! I have a few feature requests for you:
- ability to edit expander title without editing source
- ability to format title (saving eg. header tag as an attribute would go a long way
- ability to have expanders inside list (see also the bug below)
- saving expander state (cf. `open=""` attribute of `details`) 

Nice work! Before you open a PR, I suggest cleaning up some of the extra logger statements - they're quite noisy. Regarding the keyboard shortcuts, Jaap can probably point you in the right direction - reading keycodes manually and re-implementing editor logic are not the way to go. I don't know much about Gtk, but #2473 might offer a clue as to how those signals are handled.

![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/26069787/b6a6a10b-3279-4c54-a19e-b004f4ae70fb)

**Reply by mbluett88** [reply]
Strange about the displayed text you see. I don't see anything that looks like what is in your snapshot.

This is what I see:
![Screenshot from 2024-03-11 11-04-46](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/64982429/72f31073-0573-43d5-a07d-5346640175bf)

**Comment by mbluett88** [comment]
> Do you mean visually? Mayhaps you could post some screenshots of what you have and/or are working towards.

When using the InsertedObjectTypeExtension class an object buffer must be created. It is used for both display and updating the Zim page that houses that object so that any changes are saved to the associated page file. That is what those two methods are mentioned, previously, are for (i.e., making sure content changes are saved to the parent Zim page).

I have included a screenshot, but it won't demonstrate how the page saving works.
![Screenshot from 2024-01-28 13-37-13](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/64982429/93d6e080-0672-4c0a-b717-0e2c80aead79)

> As I've typed before, - "expanding/collapsing text is a pretty fundamental editor feature so the maintainability of the code is pretty paramount to its introduction."

Maintainability of plugins or even main code is always a problem. That is why the open source method is preferred (i.e., there is no guarantee having this functionality built into the main code OR having it as a plugin will not require future maintenance). There are, currently, Gtk methods used  within the main Zim code that are declared as deprecated in future releases of the Gtk library. If they are not fixed, eventually Zim will break.

> Don't worry, being on GitHub doesn't equate to being ready for primetime - take my numerous draft pull requests on this repo for example.

Good to know.

So, I would add the plugin under zim-desktop-wiki/tree/develop/zim/plugins?

Seems that if I did that and the plugin is not completely functional, that it would get released along with the next full release in a state that is less than desirable. Or, am I missing something?

Incidentally, what other projects do you work on?

**Reply by introt** [reply]
Thanks for the screenshot, the visual differentiation is visible there. While I believe this is possible to achieve via other means inside a single TextView, I do not know my way around Gtk enough to give an example - Geany uses `scintilla`, and the LLMs don't seem to be too well versed in Gtk3. <details><summary>Here's what one "AI" came up with to demonstrate the invisible=true tag attribute, which I believe is what "hidden text ranges" refers to.</summary>

Some edits were made by yours truly as the machine went for a very human off-by-one error and forgot to actually implement toggling.

```python3
import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk, Gdk

class MyTextView:
    def __init__(self):
        self.window = Gtk.Window()
        self.window.connect("destroy", Gtk.main_quit)

        self.text_view = Gtk.TextView()
        self.text_view.set_editable(True)
        self.text_view.set_justification(Gtk.Justification.FILL)
        self.text_view.set_indent(10)

        self.buffer = self.text_view.get_buffer()

        # Insert text with foldable sections
        self.buffer.set_text("This is a sample text. -> [Click\nto fold] <- Here is the folded text.")

        # Create a tag for folding
        self.folded_tag = self.buffer.create_tag("folded", invisible=True)
        self.folded = False

        # Create a button to toggle folding
        self.button = Gtk.Button(label="Toggle Folding")
        self.button.connect("clicked", self.toggle_folding)

        # Create a vertical box to hold the button and the text view
        self.box = Gtk.VBox()
        self.box.pack_start(self.button, False, False, 0)
        self.box.pack_start(self.text_view, True, True, 0)

        self.window.add(self.box)
        self.window.show_all()

    def toggle_folding(self, button):
        # Toggle the visibility of the folded sections
        start_iter = self.buffer.get_iter_at_offset(26)
        end_iter = self.buffer.get_iter_at_offset(42)
        #start_iter, end_iter = self.buffer.get_bounds()

        if self.folded:
            self.buffer.remove_tag(self.folded_tag, start_iter, end_iter)
            self.folded = False
        else:
            self.buffer.apply_tag(self.folded_tag, start_iter, end_iter)
            self.folded = True

if __name__ == "__main__":
    MyTextView()
    Gtk.main()


```
</details>

It's not great, but it does demonstrate some of the possibilities - and problems - of the approach: while hidden text doesn't take up space, the newline still does.

> Maintainability of plugins or even main code is always a problem.

Yes; what I meant especially was that working around missing _official stable plugin API_ is akin to building a _non-documented unofficial unstable API_, ie. it will make the plugin depend on internal implementation details that change between releases unannounced all the time. That's the opposite of what we'd want for foldable text which is usually a first-class feature.

> So, I would add the plugin under zim-desktop-wiki/tree/develop/zim/plugins?
>
> Seems that if I did that and the plugin is not completely functional, that it would get released along with the next full release in a state that is less than desirable. Or, am I missing something?

You are, but don't worry - this is a common misconception about open source development. On GitHub particularly, you can not add the plugin under zim-desktop-wiki/zim-desktop-wiki`/tree/develop/zim/plugins`, but you _can_ add it under mbluett88/zim-desktop-wiki`/tree/develop/zim/plugins`; this is likely what happens if you try to do the former by dragging and dropping the files into your browser window.

When you attempt to do any edits to zim-desktop-wiki/zim-desktop-wiki, GitHub will _fork_ (make a copy of) the repository for you, and your changes will only affect that fork. To incorporate your changes into the zim-desktop-wiki/zim-desktop-wiki _tree_, to be included in a future release, you will then need to open a _pull request_, also known as a PR. PRs are reviewed by eg. me, and can be modified to include any requested changes. Finally, the PR needs to be accepted by someone with the write permission to the zim-desktop-wiki/zim-desktop-wiki tree, which currently means our one-and-only BDFL Jaap.

> Incidentally, what other projects do you work on?

Currently not that many under this pseudonym - you can see some of my past projects in my profile. I did contribute a bit to FooSoft/anki-connect to make development of introt/zim2anki a bit less of a pain. Anything2Anki tools are something I'm not comfortable with publishing before the data structures are stable as changes will always break everyone's workflow which doesn't exactly attract users and contributors.

**Reply by mbluett88** [reply]
Your idea of hidden text is certainly a reasonable thing to add to Zim. However, that is meant for small sections of text. My Expander object is meant to incorporate large sections of text.

I am working on a new version that is integrated directly into the main Zim code, specifically PageView. That should permit the Zim menu formatting items to affect the text within my Expander object.

I am far from being an expert in the use of Gtk either. Learning a bit more here and there as is needed: Lots of experimentation including with the Zim code.

I sometimes use PDB to try and understand how the Zim code functions. But mainly I add 'logger' statements to the Zim code to obtain a trace. Clunky, but it gets me where I need to be.

I do my development testing inside Docker.

Although I understand object-oriented programming fairly well, I, sometimes, find it very convoluted to follow. Hence the reason for PDB. 

There are functions in Python that don't, yet, have a really good handle on. 

I have a book called "Advanced Guide to Python 3 Programming" (2023 - 2ed) by John Hunt that is by far the best I have found with regard to understanding the intricacies of Python. I have not dug into sufficiently yet.

**Reply by mbluett88** [reply]
As to Github, would you recommend that I wait and add my changes to PageView or add my original plugin method or both?

I am not sure how, even if I have a separate 'mbluett' branch, to keep things separate. I have books and tutorials on Github but have not taken the time to dig into them yet. Maybe I should start there.

**Comment by mbluett88** [comment]
Since Jaap seems adamant in having this Text Expander (or possibly better name "Folding Text" as proposed by Introt) implemented as a plugin, I have been working towards that and have succeeded with a lot of it (i.e, formatting, bullets, links, etc.). However, part of the Links effort requires having the "Remove Link" menu item and the "Edit Link or Object" menu item enabled for use within the Expander TextView. This enabling occurs in PageView.do_mark_set().

All of my efforts to enable these menu items have failed, unless of course, I just force them to be permanently enabled, which is not what was originally intended.

The only solution I can see is to have a reference to the pageview instance available within the Expander plugin. 

I have another implementation where these menu items work properly but it is not based on a plugin and it, of course, makes use of the pageview instance.

If someone has a suggestion of how to make these menuitems function the way they were intended to function, without making use of the pageview instance, I am all ears.

**Comment by jaap-karssenberg** [comment]
Here the pageview is listening to a signal from the textbuffer. To make the
same logic work inside an embedded object, the object should also emit the
relevant signal(s) and the pageview should be able to handle multiple text
buffers.

I'm fine with adding that logic in the core classes provided that we make
this signal interface generic enough such that other plugins can also use
it.

The signal-listener design pattern is a good way to approach this
interface. It allows the expander object to trigger the pageview without
needing a direct reference.

Does this help?

Jaap



Op ma 6 mei 2024 22:28 schreef Mike ***@***.***>:

**Reply by mbluett88** [reply]
Sorry, Jaap, but it does not help. I have tried implementing various signals only to be stymied by Zim code and a limitation in Python.

For example, in  PageView: _connect_focus_event(self), this code exists:
`	window = self.get_toplevel()
	if window and window != self:
		window.connect('set-focus', set_actiongroup_sensitive)
`
This is executed before any plugins are instantiated. So, I can't make the connect here.

I tried implementing a connect() for 'set-focus' to make use of set_actiongroup_sensitive() but there is no way to do this from Expander.

Using something like PageView.set_actiongroup_sensitive() does not work because of 'self'.

The only way to make this work is with an actual reference to the pageview instance.

**Reply by introt** [reply]
> I tried implementing a connect() for 'set-focus'

I think you should `emit('set-focus', widget)` to trigger `set_actiongroup_sensitive()`, not `connect()` to it - connect connects the signal to its handler, see signals.py for some docs.

**Reply by mbluett88** [reply]
That is right 'emit' is necessary. However, in order for the 'emit' to work you first have to 'connect' the signal to a textview. In this case, my Expander textview. And to do that the only way you can access the set_actiongroup() method is by either using PageView or by using the instance of pageview.

PageView does not work because it is not a real object instance and therefore there is no way to access attributes of the actual pageview object instance. This means I would need to use the actual pageview instance that has previously been instantiated. I see no other way to access those attributes (e.g., 'remove link').

**Reply by mbluett88** [reply]
Actually, the "Remove Link" and "Edit Link or Object" now work without making use of the pageview instance. I haven't tried this using the Plugin version. But, it works well within the more integrated version. 

Basically, the integrated version is dependent on InsertedObjectAnchor instead of InsertedObjectTypeExtension (in the Plugin version).

**Comment by mbluett88** [comment]
I have most of the bugs worked out and the Integrated version is in pretty good shape.

It permits applying all format types. All inserts work. Copying and pasting between buffers works. Cut works, Ctrl-Z (undo) Shift-Ctrl-Z (redo) works.

I have tested using multiple Expander widgets as well.

The only thing that doesn't work as of yet is search across all Expander widgets. I view this as important to the functioning of this new Expander facility. But, I suppose it can wait a bit.

Just discovered one other thing that doesn't work. Can't use SourceView within an Expander buffer space.

Jaap, if you feel OK with this, I can upload this version and do a Pull Request.

I could possibly add the things, I have added to this Integrated version, to the Plugin version, but I am getting kind of tired of switching back and forth between the two versions. I have spent a good deal of time getting this thing working. So, I would prefer to release for inclusion soon.

If anyone would care to hop in and test or help perhaps with the search part that would be very welcome.

**Comment by mbluett88** [comment]
I am trying to get the right-click context menu (that is available in the PageView context) to work with my Expander object.

In PageView, there is a connect like this:
       self.textview.connect_object('populate-popup', PageView.do_populate_popup, self)

I want to make the same connect in my Expander object. But, I can't do that because the 'self' ends up being ExpanderWidget instead of PageView.

I need access to the do_populate_popup() method of the actual 'pageview' instance.

There must be some way to make this connection but I am not aware of how to make this work.

Any suggestions?

**Reply by mbluett88** [reply]
I think the big problem with trying to get access to menu functionality that PageView uses is that Zim was designed around the idea that PageView would be the only object type ever used, to manage pages.

I think what should be the case is that all menu items should be handled within a Menu class. In that way, PageView could make use of the related menu items as well as other object classes.

I noticed that there is a comment in PageView._default_do_populate_popup() that talks about the need for a deepcopy. If the menu items were incorporated into a Menu class, this likely would not be needed. So, do we need a redesign?
